### PR TITLE
improved mode_drip

### PR DESCRIPTION
**Pull request by ewoudwijma**
- Support palettes
- One drop after the other
- Not fall random but fall if swelled (swell has random component)
- Add Swelling slider (not combined with gravity)

### CODE REVIEW

File: wled00/FX.cpp [hunk 1]
```diff
@@ -3385,58 +3391,57 @@ uint16_t mode_drip(void)
   struct virtualStrip {
     static void runStrip(uint16_t stripNr, Spark* drops) {
 
-      uint8_t numDrops = 1 + (SEGMENT.intensity >> 6); // 255>>6 = 3
+      uint8_t numDrops = map(SEGMENT.intensity, 0, 255, 1, 4);
```
Comment (blazoncek) [inline wled00/FX.cpp:3394]: Cleaner but uses more instructions. Inefficient.

File: wled00/FX.cpp [hunk 2]
```diff
@@ -3385,58 +3391,57 @@ uint16_t mode_drip(void)
...
 
-      float gravity = -0.0005 - (SEGMENT.speed/50000.0);
+      float gravity = -0.0005 - (SEGMENT.speed/25000.0); //increased max speed 50000->25000
```
Comment (blazoncek) [inline wled00/FX.cpp:3396]: Did you test on short segments 10 pixels and/or long segments 300+ pixels?
May behave too fast if not tested.

File: wled00/FX.cpp [hunk 3]
```diff
@@ -3385,58 +3391,57 @@ uint16_t mode_drip(void)
...
 
       for (int j=0;j<numDrops;j++) {
-        if (drops[j].colIndex == 0) { //init
+        if (drops[j].colIndex == 0) { // drop state (0 init, 1 forming, 2 falling, 5 bouncing)
           drops[j].pos = SEGLEN-1;    // start at end
           drops[j].vel = 0;           // speed
-          drops[j].col = sourcedrop;  // brightness
-          drops[j].colIndex = 1;      // drop state (0 init, 1 forming, 2 falling, 5 bouncing)
+          drops[j].col = 0;           // brightness, start with 0
+          drops[j].velX = SEGMENT.color_from_palette(random8(), false, PALETTE_SOLID_WRAP, 0); // random color
```
Comment (blazoncek) [inline wled00/FX.cpp:3405]: Assign velX to a uint8_t variable to avoid unnecessary float to byte conversions further down.

File: wled00/FX.cpp [hunk 4]
```diff
@@ -3385,58 +3391,57 @@ uint16_t mode_drip(void)
...
             drops[j].col=255;
+            drops[j].colIndex=2;               // fall
           }
+          //draw forming drop (lower the brightness so forming is more visible)
+          SEGMENT.setPixelColor(indexToVStrip(uint16_t(drops[j].pos), stripNr), color_blend(BLACK, drops[j].velX, drops[j].col/8));
+
+          drops[j].col += mapf(SEGMENT.custom1, 0, 255, 0.01, 2.5) * random8()/128; // swelling with randomness
         }
-        if (drops[j].colIndex > 1) {           // falling
+        else if (drops[j].colIndex > 1) {      // falling
```
Comment (blazoncek) [inline wled00/FX.cpp:3421]: I prefered old way with multiple bounces.

File: wled00/FX.cpp [hunk 5]
```diff
@@ -3385,58 +3391,57 @@ uint16_t mode_drip(void)
...
-          drops[j].col = sourcedrop;  // brightness
-          drops[j].colIndex = 1;      // drop state (0 init, 1 forming, 2 falling, 5 bouncing)
+          drops[j].col = 0;           // brightness, start with 0
+          drops[j].velX = SEGMENT.color_from_palette(random8(), false, PALETTE_SOLID_WRAP, 0); // random color
+          if (j==0) drops[j].colIndex = 1; // only first drop starts forming
         }
+        // if drop is falling, form the next drop
+        if (j>0 && drops[j-1].colIndex > 1 && drops[j].colIndex == 0) drops[j].colIndex = 1;
 
-        SEGMENT.setPixelColor(indexToVStrip(SEGLEN-1, stripNr), color_blend(BLACK,SEGCOLOR(0), sourcedrop));// water source
```
Comment (blazoncek) [inline wled00/FX.cpp:3402]: Watering was essential visual element.

File: wled00/FX.cpp [hunk 6]
```diff
@@ -3385,58 +3391,57 @@ uint16_t mode_drip(void)
...
+              SEGMENT.setPixelColor(indexToVStrip(pos, stripNr), color_blend(BLACK, drops[j].velX, drops[j].col/i)); //spread pixel with fade while falling
             }
 
             if (drops[j].colIndex > 2) {       // during bounce, some water is on the floor
-              SEGMENT.setPixelColor(indexToVStrip(0, stripNr), color_blend(SEGCOLOR(0),BLACK,drops[j].col));
+              SEGMENT.setPixelColor(indexToVStrip(0, stripNr), color_blend(drops[j].velX, BLACK, drops[j].col));
             }
           } else {                             // we hit bottom
-            if (drops[j].colIndex > 2) {       // already hit once, so back to forming
+            if (drops[j].colIndex == 5) {      // already hit once, so back to forming
```
Comment (blazoncek) [inline wled00/FX.cpp:3436]: See above.

**Review by blazoncek — CHANGES_REQUESTED** [review]
I am not sure adding palette color selection benefits the effect though I am not against it.
I do not understand the *swelling* behaviour (it seems not an upgrade of watering) but *watering* was essential element of the effect and I would preferred it remained intact.

**Comment by blazoncek** [comment]
Let's ask @Aircoookie what's his opinion.

File: wled00/FX.cpp (hunk 1 above)
Comment (ewoudwijma) [reply wled00/FX.cpp]: Agree / reversed

File: wled00/FX.cpp (hunk 2 above)
Comment (ewoudwijma) [reply wled00/FX.cpp]: Yes, it is a range so all values possible

File: wled00/FX.cpp (hunk 3 above)
Comment (ewoudwijma) [reply wled00/FX.cpp:3400]: done

File: wled00/FX.cpp (hunk 4 above)
Comment (ewoudwijma) [reply wled00/FX.cpp]: Agree / reversed

File: wled00/FX.cpp (hunk 5 above)
Comment (ewoudwijma) [reply wled00/FX.cpp:3402]: Agree / reversed

File: wled00/FX.cpp (hunk 6 above)
Comment (ewoudwijma) [reply wled00/FX.cpp]: n/a

**Comment by blazoncek** [comment]
Tested this and must say that watering does not look as original to my eye. It swells too fast.
If you set Fall rate slider to 0 no drops fall.

**Comment by ewoudwijma** [comment]
swelling should be the same but I will check. 
Will also check fall rate slider

**Comment by blazoncek** [comment]
Not going merge this.

**Pull request closed without merging**